repositoryref.go

```diff
@@ -0,0 +1,258 @@
+	return o.SubOrganizations
+}
+
+// organizationString returns the organizations of an OrganizationRef slash-separated
+func organizationString(o OrganizationRef) string {
+	orgParts := append([]string{o.GetOrganization()}, o.GetSubOrganizations()...)
+	return strings.Join(orgParts, "/")
+}
+
+// String returns the HTTPS URL to access the Organization
+func (o OrganizationInfo) String() string {
+	return fmt.Sprintf("https://%s/%s", o.GetDomain(), organizationString(o))
+}
+
+// RefIsEmpty returns true if all the parts of the given OrganizationInfo are empty, otherwise false
+func (o OrganizationInfo) RefIsEmpty() bool {
+	return len(o.Domain) == 0 && len(o.Organization) == 0 && len(o.SubOrganizations) == 0
+}
+
+// validateOrganizationInfoCreate validates its own field into a given error list
+func (o OrganizationInfo) validateOrganizationInfoCreate(errs *validationErrorList) {
+	// Require the Domain and Organization to be set
+	if len(o.Domain) == 0 {
+		errs.Required("Domain")
+	}
+	if len(o.Organization) == 0 {
+		errs.Required("Organization")
+	}
+}
+
+// RepositoryInfo is an implementation of RepositoryRef
+type RepositoryInfo struct {
+	// RepositoryInfo embeds everything in OrganizationInfo inline
+	OrganizationInfo `json:",inline"`
+
+	// Name specifies the Git repository name. This field is URL-friendly,
+	// e.g. "kubernetes" or "cluster-api-provider-aws"
+	// +required
+	RepositoryName string `json:"repositoryName"`
+}
+
+// RepositoryInfo implements the RepositoryRef interface
+var _ RepositoryRef = RepositoryInfo{}
+
+// GetRepository returns the name of the repository
+func (r RepositoryInfo) GetRepository() string {
+	return r.RepositoryName
+}
+
+// String returns the HTTPS URL to access the Repository
+func (r RepositoryInfo) String() string {
+	return fmt.Sprintf("%s/%s", r.OrganizationInfo.String(), r.GetRepository())
+}
+
+// RefIsEmpty returns true if all the parts of the given RepositoryInfo are empty, otherwise false
+func (r RepositoryInfo) RefIsEmpty() bool {
+	return r.OrganizationInfo.RefIsEmpty() && len(r.RepositoryName) == 0
+}
+
+// validateRepositoryInfoCreate validates its own field into a given error list
+func (r RepositoryInfo) validateRepositoryInfoCreate(errs *validationErrorList) {
+	// First, validate the embedded OrganizationInfo
+	r.validateOrganizationInfoCreate(errs)
+	// Require RepositoryName to be set
+	if len(r.RepositoryName) == 0 {
+		errs.Required("RepositoryName")
+	}
+}
+
+// GetCloneURL gets the clone URL for the specified transport type
+func (r RepositoryInfo) GetCloneURL(transport TransportType) string {
+	return GetCloneURL(r, transport)
+}
+
+// GetCloneURL returns the URL to clone a repository for a given transport type. If the given
+// TransportType isn't known an empty string is returned.
+func GetCloneURL(rs RepositoryRef, transport TransportType) string {
+	switch transport {
+	case TransportTypeHTTPS:
+		return fmt.Sprintf("%s.git", rs.String())
```

Comment:
What if the URL contains `.git`, we should not append it that case.

Reply:
Thanks for catching that.
I'm now excluding any possible `.git` suffix when parsing the URL to make `rs.String()` always return the HTTPS URL for the web page (without the .git suffix to make it safe to append)